plugins/rbac-backend/src/service/policies-rest-api.ts

```diff
@@ -240,71 +406,122 @@ export class PolicesServer {
+        throw new InputError(`'newRole' object must be present`); // 400
       }
 
-      oldPolicyRaw.entityReference = entityRef;
-      let err = validatePolicy(oldPolicyRaw);
+      oldRoleRaw.roleName = roleEntityRef;
+      let err = validateRole(oldRoleRaw);
       if (err) {
         throw new InputError( // 400
-          `Invalid old policy object. Cause: ${err.message}`,
+          `Invalid old role object. Cause: ${err.message}`,
         );
       }
-      newPolicyRaw.entityReference = entityRef;
-      err = validatePolicy(newPolicyRaw);
+      newRoleRaw.roleName = roleEntityRef;
+      err = validateRole(newRoleRaw);
       if (err) {
         throw new InputError( // 400
-          `Invalid new policy object. Cause: ${err.message}`,
+          `Invalid new role object. Cause: ${err.message}`,
         );
       }
 
-      const oldPolicy = this.transformPolicyToArray(oldPolicyRaw);
-      const newPolicy = this.transformPolicyToArray(newPolicyRaw);
+      const oldRole = this.transformRoleToArray(oldRoleRaw);
+      const newRole = this.transformRoleToArray(newRoleRaw);
+
+      if (isEqual(oldRole, newRole)) {
+        response.status(204).end();
+        return;
+      }
 
-      if (await this.enforcer.hasPolicy(...newPolicy)) {
-        if (isEqual(oldPolicy, newPolicy)) {
-          resp.status(204).end();
-          return;
+      for (const role of newRole) {
+        const hasRole = oldRole.some(element => {
+          return isEqual(element, role);
+        });
+        // if the role is already part of old role and is a grouping policy we want to skip returning a conflict error
+        // to allow for other roles to be checked and added
+        if ((await this.enforcer.hasGroupingPolicy(...role)) && !hasRole) {
+          throw new ConflictError(); // 409
         }
-        throw new ConflictError(); // 409
       }
 
-      if (!(await this.enforcer.hasPolicy(...oldPolicy))) {
-        throw new NotFoundError(); // 404
+      for (const role of oldRole) {
+        if (!(await this.enforcer.hasGroupingPolicy(...role))) {
+          throw new NotFoundError(); // 404
+        }
       }
 
-      // enforcer.updatePolicy(oldPolicyPermission, newPolicyPermission) was not implemented
+      // enforcer.updateGroupingPolicy(oldRole, newRole) was not implemented
       // for ORMTypeAdapter.
       // So, let's compensate this combination delete + create.
-      const isRemoved = await this.enforcer.removePolicy(...oldPolicy);
+      const isRemoved = await this.enforcer.removeGroupingPolicies(oldRole);
       if (!isRemoved) {
         throw new Error('Unexpected error'); // 500
       }
 
-      const isAdded = await this.enforcer.addPolicy(...newPolicy);
+      const isAdded = await this.enforcer.addGroupingPolicies(newRole);
       if (!isAdded) {
         throw new Error('Unexpected error');
       }
 
-      resp.status(200).end();
+      response.status(200).end();
     });
 
+    router.delete(
```

Comment:
I think delete method should delete role, corresponding permission policies and user, group references. What do you think, @PatAKnight ?

Reply:
Right now, this will only delete a single group / user from a role. So it doesn't remove the role completely if there are other users / groups assigned to it. I figured this approach would be a good starting point.

But we can add the option to delete the entire role (all users / groups attached to a role) if there are no query params listed. Then we can go through and remove all policies that are associated with that role.

Reply:
I have updated it to now delete an entire role. I still left in the functionality to delete a single user / group from a role and will just document the differences for now.